src/components/Map/index.tsx

```diff
@@ -0,0 +1,190 @@
+import * as React from 'react';
+import { GoogleMap, LoadScript, Marker } from '@react-google-maps/api';
+import { DefaultProps } from '../../common-types';
+import { Box } from '../Box';
+
+export interface UIMarker {
+	id: string | number;
+	lat: number;
+	lng: number;
+	title?: string;
+	icon?: string;
+	click?: () => void;
+}
+
+// Coordinates of Seattle, avoid having the map centered in the middle of the sea
+const DEFAULT_LATLNG = {
+	lat: 47.6062,
+	lng: -122.3321,
+};
+const DEFAULT_ZOOM_LEVEL = 3;
+const DETAIL_ZOOM_LEVEL = 12;
+
+const defaultMapOptions: google.maps.MapOptions = {
+	controlSize: 32,
+	scrollwheel: false,
+	disableDefaultUI: true,
+	zoomControl: true,
+	mapTypeControl: true,
+	mapTypeControlOptions: {
+		mapTypeIds: ['roadmap', 'hybrid'],
+	},
+	styles: [
+		{
+			elementType: 'all',
+			featureType: 'all',
+			stylers: [{ saturation: -50 }],
+		},
+		{
+			elementType: 'all',
+			featureType: 'administrative.country',
+			stylers: [{ gamma: 2 }],
+		},
+		{
+			elementType: 'all',
+			featureType: 'administrative.locality',
+			stylers: [{ gamma: 1.5 }],
+		},
+	],
+};
+
+// Return map bounds based on list of markers
+const getMapBounds = (markers: UIMarker[]) => {
+	const bounds = new google.maps.LatLngBounds();
+
+	markers.forEach((marker) => {
+		bounds.extend(
+			new google.maps.LatLng(
+				isNaN(marker.lat) ? 0 : marker.lat,
+				isNaN(marker.lng) ? 0 : marker.lng,
+			),
+		);
+	});
+	return bounds;
+};
+
+// Fit map to its bounds after the api is loaded
+const apiIsLoaded = (map: google.maps.Map, markers: UIMarker[]) => {
+	if (!markers.length) {
+		map.setCenter(DEFAULT_LATLNG);
+		map.setZoom(DEFAULT_ZOOM_LEVEL);
+	} else {
+		const bounds = getMapBounds(markers);
+		map.fitBounds(bounds);
+	}
+
+	// These options depend on the global "google" object, so we set it once the API has loaded.
+	map.setOptions({
+		zoomControlOptions: {
+			position: google.maps.ControlPosition.TOP_RIGHT,
+			style: google.maps.ZoomControlStyle.SMALL,
+		},
+	});
+
+	// If we run `setZoom` right after `fitBounds` the map won't refresh. With this we first wait for the map to be idle (from fitBounds), and then set the zoom level.
+	const listener = google.maps.event.addListenerOnce(map, 'idle', () => {
+		// Don't allow to zoom closer than the defailt detail zoom level on initial load.
+		if (map.getZoom() > DETAIL_ZOOM_LEVEL) {
+			map.setZoom(DETAIL_ZOOM_LEVEL);
+		}
+
+		google.maps.event.removeListener(listener);
+	});
+};
+
+const getFieldFromMap = <T extends any, U extends any>(
+	entry: T,
+	fieldName: keyof MapProps<T>['dataMap'],
+	dataMap: MapProps<T>['dataMap'],
+) => {
+	const field = dataMap[fieldName].find((x) => !!entry[x]);
+	return field ? (entry[field] as U) : undefined;
+};
+
+const Map = <T extends any>({
+	className,
+	data = [],
+	dataMap,
+	getIcon,
+	onItemClick,
+	apiKey,
+	mapClick,
+}: MapProps<T>) => {
+	const markers = React.useMemo(
```

Comment:
Why not just loop below? I don't see any of these operations as complex that they should be memoized.

Reply:
@iamsolankiamit if there are 10000 items for example, this helps with performance.